File: prime-number/node/np.js
const calc = (args) => {

    const primeNumbers = []

    const isPrimeNumber = (n) => {
        for (let i = 0; i < primeNumbers.length; i++) {
            const pn = primeNumbers[i]
            if (pn * pn > n)
                return true
            if (n % pn  == 0)
                return false
        }
        return true
    }

    for (let n = 2; n <= args.upperLimit; n++) {
        if (isPrimeNumber(n))
            primeNumbers.push(n)
    }

    console.log("Count: " + primeNumbers.length)
    if (args.show) {
        for (let i = 0; i < primeNumbers.length; i++)
            console.log(primeNumbers[i])
    }
}

const parseArgs = (argv) => {
    const args = {
        upperLimit: 0,
        show: false,
    }

    let nextParser = null;
    for (let i = 0; i < argv.length; i++) {
        const arg = argv[i]
        if (nextParser) {
            nextParser = nextParser(arg)
            continue
        }
        if (arg == "-u" || arg == "--upper-limit") {
            nextParser = (arg) => {
                args.upperLimit = Number(arg)
                return null
            }
        }
        else if (arg == "-s" || arg ==  "--show") {
            args.show = true
        }
    }
    return args
}

const start = () => {
    const args = parseArgs(process.argv)
    console.log("UpperLimit: " + args.upperLimit)
    calc(args)
}

start()

File: prime-number/kotlin/src/main/kotlin/np.kt
data class Args(var upper: Int = 100, var showResult: Boolean = false,
                var varTypeLong: Boolean = false)

fun <T> showResult(numbers: Collection<T>, args: Args) {
    println(numbers.size)
    if (args.showResult)
        println(numbers)
}

class PrimeNumberInt {
    val primeNumbers = mutableListOf(2)

    fun isPrimeNumber(n: Int): Boolean {
        for (pn in primeNumbers) {
            if (pn * pn > n)
                return true
            if (n % pn == 0)
                return false
        }
        return true
    }

    fun run(args: Args) {
        var n = 3
        while (n < args.upper) {
            if (isPrimeNumber(n))
                primeNumbers.add(n)
            n += 2
        }
        showResult(primeNumbers, args)
    }
}

class PrimeNumberLong {
    val primeNumbers = mutableListOf(2L)

    fun isPrimeNumber(n: Long): Boolean {
        for (pn in primeNumbers) {
            if (pn * pn > n)
                return true
            if (n % pn == 0L)
                return false
        }
        return true
    }

    fun run(args: Args) {
        var n = 3L
        while (n < args.upper) {
            if (isPrimeNumber(n))
                primeNumbers.add(n)
            n += 2L
        }
        showResult(primeNumbers, args)
    }
}

fun parseArgs(argArr: Array<String>): Args {
    var idx = 0
    val args = Args()

    fun isEnd(): Boolean {
        return idx >= argArr.size
    }

    fun loadNext() {
        idx++
    }

    fun ref(): String {
        return argArr[idx]
    }

    fun refAsInt(): Int {
        return Integer.parseInt(ref())
    }

    while (!isEnd()) {
        when (ref()) {
            "-u", "--upper-limit" -> {
                loadNext()
                args.upper = refAsInt()
            }
            "-s", "--show" -> {
                args.showResult = true
            }
            "--long" -> {
                args.varTypeLong = true
            }
        }
        loadNext()
    }
    return args
}

fun main(argArr: Array<String>) {
    val args = parseArgs(argArr)
    println("Upper bound: " + args.upper)
    if (args.varTypeLong) {
        PrimeNumberLong().run(args)
    } else {
        PrimeNumberInt().run(args)
    }
}
